packages/contracts-bedrock/src/L2/CrossL2Inbox.sol

```diff
@@ -0,0 +1,157 @@
+    bytes32 public constant TIMESTAMP_SLOT = 0x2e148a404a50bb94820b576997fd6450117132387be615e460fa8c5e11777e02;
+
+    /// @notice Transient storage slot that the chainId for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.chainid")) - 1)
+    bytes32 public constant CHAINID_SLOT = 0x6e0446e8b5098b8c8193f964f1b567ec3a2bdaeba33d36acb85c1f1d3f92d313;
+
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Enforces that cross domain message sender and source are set. Reverts if not.
+    ///         This is leveraged to differentiate between 0 and nil at tstorage slots.
+    modifier notEntered() {
+        assembly {
+            if eq(tload(ENTERED_SLOT), 0) {
+                mstore(0x00, 0xbca35af6) // 0xbca35af6 is the 4-byte selector of "NotEntered()"
+                revert(0x1C, 0x04)
+            }
+        }
+        _;
+    }
+
+    /// @notice Returns the origin address of the Identifier. If not entered, reverts.
+    /// @return _origin The origin address of the Identifier.
+    function origin() external view notEntered returns (address _origin) {
+        assembly {
+            _origin := tload(ORIGIN_SLOT)
+        }
+    }
+
+    /// @notice Returns the block number of the Identifier. If not entered, reverts.
+    /// @return _blocknumber The block number of the Identifier.
+    function blocknumber() external view notEntered returns (uint256 _blocknumber) {
+        assembly {
+            _blocknumber := tload(BLOCKNUMBER_SLOT)
+        }
+    }
+
+    /// @notice Returns the log index of the Identifier. If not entered, reverts.
+    /// @return _logIndex The log index of the Identifier.
+    function logIndex() external view notEntered returns (uint256 _logIndex) {
+        assembly {
+            _logIndex := tload(LOG_INDEX_SLOT)
+        }
+    }
+
+    /// @notice Returns the timestamp of the Identifier. If not entered, reverts.
+    /// @return _timestamp The timestamp of the Identifier.
+    function timestamp() external view notEntered returns (uint256 _timestamp) {
+        assembly {
+            _timestamp := tload(TIMESTAMP_SLOT)
+        }
+    }
+
+    /// @notice Returns the chain ID of the Identifier. If not entered, reverts.
+    /// @return _chainId The chain ID of the Identifier.
+    function chainId() external view notEntered returns (uint256 _chainId) {
+        assembly {
+            _chainId := tload(CHAINID_SLOT)
+        }
+    }
+
+    /// @notice Executes a cross chain message on the destination chain.
+    /// @param _id An Identifier pointing to the initiating message.
+    /// @param _target Account that is called with _msg.
+    /// @param _msg The message payload, matching the initiating message.
+    function executeMessage(Identifier calldata _id, address _target, bytes memory _msg) external payable {
+        require(msg.sender == tx.origin, "CrossL2Inbox: not EOA sender");
```

Comment:
In a follow up PR:
- remove only eoa invariant
- migrate to custom errors instead of `require` (right now its mixed)
- use transient context modifier to allow reentrency
- test reentrency works with a multicall